Gatsby source plugin (#1297)
SenseNet/sn-client: packages/gatsby-source-sensenet/README.md

# gatsby-source-sensenet

Source plugin for pulling data from [sensenet](https://sensenet.com) into Gatsby websites.

sensenet is an open-source headless content management system (CMS) built mainly for developers and development companies.
It is a content repository where you can store all your content and reach it through APIs. It is a solid base for your custom solutions, offering an enterprise-grade security and permission system, versioning, dynamic content types and even more.

> An example site for using this plugin is at [sn-blog-gatsby](https://github.com/SenseNet/sn-client/tree/develop/examples/sn-blog-gatsby)

Reviewer: ```suggestion
An example of how to use this plugin is at [sn-blog-gatsby](https://github.com/SenseNet/sn-client/tree/develop/examples/sn-blog-gatsby)
```

## How to install

```shell
npm install gatsby-source-sensenet
```

## Available options

> You need a way to pass environment variables to the build process, so secrets and other secured data should not commit to source control.

Reviewer: ```suggestion
You have to pass environment variables to the build process, so secrets and other secured data will not commited along the source code.
```

```javascript
// In your gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: `gatsby-source-sensenet`,
      options: {
        host: '<YOUR REPOSITORY URL>',
        path: '<RELATIVE PATH TO YOUR CONTENTS>',
        oDataOptions: '<ODATA OPTIONS>'
        accessToken: '<ACCESS TOKEN FOR AUTHENTICATION>'
        level: '<NUMBER OF LEVELS TO READ CONTENTS>'
      },
    },
  ],
}
```

**`host`** [string][required]

The url of your repository, e.g.: 'https://dev.demo.sensenet.com'

**`path`** [string][optional] [default: '/Root/Content']

This is the root path of the container where your stuff are located in your repository.

> The default value is 'Root/Content' but if you don't want to load all the content from your repository you can reduce the number of content request by specifying the container (e.g.: folder) that contains all the contents you will need to create a site.

Reviewer: ```suggestion
The default value is 'Root/Content' but if you don't want to load all the content from your repository you can reduce the number of content request by specifying a container (e.g.: folder, workspace or library) that contains the contents you need to build up your static site.
```

**`oDataOptions`** [[ODataParams](https://github.com/SenseNet/sn-client/blob/28bcd4f0cbf8f366ba8afa33f839c26959c78c4e/packages/sn-client-core/src/Models/ODataParams.ts#L31)][optional] [default:

```json
{
  "enableautofilters": false,
  "enablelifespanfilter": false,
  "inlinecount": "allpages",
  "metadata": "no",
  "select": ["DisplayName", "Description", "Icon"],
  "top": 10000
}
```

]

> Query options are query string parameters a client may specify to control the amount and order of the data that a service returns for the resource identified by the URI. In sensenet there're two types of query options available OData System Query Options and custom sensenet query options. The OData standard query options' names are prefixed with a "$" character, sensenet query options should be used without a prefix.

Reviewer: ```suggestion
Query options are query string parameters that the client may specify controling the amount and order of the data the service returns. In sensenet there're two types of query options available OData System Query Options and custom sensenet query options. The OData standard query options' names are prefixed with a "$" character, sensenet query options should be used without a prefix.
```

| $format                                                                                                | specifies that a response to the request MUST use the media type specified by the query option (Atom and xml formats are not implemented yet in sensenet)                            |
| [$inlinecount](https://docs.sensenet.com/api-docs/basic-concepts/02-collection#inlinecountqueryoption) | controls the `__count` property that can be found in every collection response                                                                                                       |
| [query](https://docs.sensenet.com/api-docs/basic-concepts/05-search-filter#queryoption)                | filter the result collection using sensenet Content Query                                                                                                                            |
| [metadata](https://docs.sensenet.com/api-docs/basic-concepts/06-metadata)                              | controls the metadata content in output entities                                                                                                                                     |

**`level`** [number][optional]

> The value of the level option is either a positive integer to specify the number of levels to read contents in the tree.

Reviewer: ```suggestion
The value of the level option should be a positive whole number (integer) to specify the number of levels from which you want to query content.
```

> If it is not specified then the plugin will read till a child content is exists.

Reviewer: ```suggestion
If it is not specified, the plugin will read all levels until it finds a child content.
```

**`accessToken`** [string | Function][required]

accessToken can be a string or a function as well, it is for authentication purposes.
If you would like to generate accesstoken programatically you can use _codeLogin_ function from [sn-authentication-oidc-react](https://github.com/SenseNet/sn-client/tree/develop/packages/sn-authentication-oidc-react) package.

```shell
npm install @sensenet/authentication-oidc-react
```

Example:

```javascript
// In your gatsby-config.js
const fetch = require('node-fetch')
const { codeLogin } = require('@sensenet/authentication-oidc-react')
const { configuration } = require('./configuration')

module.exports = {
  plugins: [
    {
      resolve: `gatsby-source-sensenet`,
      options: {
        host: '<YOUR REPOSITORY URL>',
        path: '<RELATIVE PATH TO YOUR CONTENTS>',
        oDataOptions: '<ODATA OPTIONS>'
        accessToken: async () => {
  const authData = await codeLogin({ ...configuration, fetchMethod: fetch })
  return authData.access_token
}
      },
    },
  ],
}

```

> You can overwrite fetch method according to your own needs. In this example [node-fetch](https://www.npmjs.com/package/node-fetch) is used, you have to install it before using it:

Reviewer: ```suggestion
You can overwrite fetch method according to your own needs. In this particlar example we are using [node-fetch](https://www.npmjs.com/package/node-fetch), so you have to install it first:
```

```shell
npm install node-fetch
```

Configuration should be something like this:

```javascript

> // Create a new file in root folder with the name of configuration.js

Reviewer: ```suggestion
// Create a new file in root folder with the name configuration.js
```

exports.repositoryUrl = '<YOUR REPOSITORY URL>'

exports.configuration = {
  clientId: process.env.GATSBY_REACT_APP_CLIENT_ID || '',
  clientSecret: process.env.GATSBY_REACT_APP_CLIENT_SECRET || '',
  identityServerUrl: '<YOUR IDENTITY SERVER URL>',
}
```

**GATSBY_REACT_APP_CLIENT_ID** and **GATSBY_REACT_APP_CLIENT_SECRET** environmental variables should be defined.

> You can easily store it in .env files by doing the following:

Reviewer: ```suggestion
You can easily store them in .env files by doing the following:
```

```javascript
// In your .env file
GATSBY_REACT_APP_CLIENT_ID=<YOUR CLIENT ID>
GATSBY_REACT_APP_CLIENT_SECRET=<YOUR SECRET>

```

### Where can you get the missing information?

There are two ways to get your client_id and client_secret:

> From your [profile site](https://docs.sensenet.com/concepts/basics/06-authentication-secrets) and from [admin-ui](https://docs.sensenet.com/guides/settings/api-and-security) as well. Here you can check your repository url and identity server url, too.

Reviewer: ```suggestion
You can find them on your [snaas user profile](https://docs.sensenet.com/concepts/basics/06-authentication-secrets) or on the [admin-ui](https://docs.sensenet.com/guides/settings/api-and-security) logged-in to the repository as well. Also here can be found the repository url and url of the identity server.
```

## When do I use this plugin?

sensenet is the single hub for all your content packed with enterprise grade features. In sensenet everything is a content (blog posts, files, users, roles, comments, etc.) delivered the same way through the API, making it super easy to work with any type of data. With sensenet's and Gatsby's flexibility and power you can build any kind of app or website you need.

> For more details check [this site](https://docs.sensenet.com/integrations//gatsby/using-gatsby-with-sensenet).

Reviewer: ```suggestion
For more details check [this guide](https://docs.sensenet.com/integrations//gatsby/using-gatsby-with-sensenet).
```

## Examples of usage

```javascript
// In your gatsby-config.js
const fetch = require('node-fetch')
const { codeLogin } = require('@sensenet/authentication-oidc-react')
const { configuration } = require('./configuration')

module.exports = {
  plugins: [
    {
      resolve: `gatsby-source-sensenet`,
      options: {
        host: 'https://dev.demo.sensenet.com',
        path: '/Root/Content/SampleWorkspace/Blog',
        oDataOptions: {
          select: 'all',
          expand: ['LeadImage'],
          metadata: 'no',
        },
        accessToken: async () => {
          const authData = await codeLogin({ ...configuration, fetchMethod: fetch })
          return authData.access_token
        },
      },
    },
  ],
}
```

## How to query for data

> gatsby-source-sensenet creates gatsby node from all the contents that return from the query. The _internal.type_ of all nodes consists of the "sensenet" prefix and the original type of the content.

Reviewer: ```suggestion
gatsby-source-sensenet creates gatsby nodes from all the contents returned by a query. The _internal.type_ of all nodes consists of the "sensenet" prefix and the original type of the content.
```

If you have a "BlogPost" type then all the content created from this type will be a gatsby node whose _internal.type_ will be: "sensenetBlogPost".

You might query for all of a type of node e.g.: sensenetBlogPost:

```graphql
  query MyQuery {
    allSensenetBlogPost {
      edges {
        node {
          Name
          DisplayName
        }
      }
    }
```

> To query for a single blogpost with the name '2021-02-23-how-we-post':

Reviewer: ```suggestion
To query for a specific blogpost with the name '2021-02-23-how-we-post':
```

  const { createPage } = actions

  const allSensenetBlogPost = await graphql(`
    {
      allSensenetBlogPost(limit: 1000) {
        edges {
          node {
            fields {
              slug
            }
          }
        }
      }
    }
  `)

  if (allSensenetBlogPost.errors) {
    console.error(allSensenetBlogPost.errors)
    throw new Error(allSensenetBlogPost.errors)
  }

  const blogPostTemplate = path.resolve('<YOUR TEMPLATE>')

  allSensenetBlogPost.data.allSensenetBlogPost.edges.forEach(({ node }) => {
    const { slug } = node.fields

    createPage({
      path: `/${slug}/`,
      component: blogPostTemplate,
      context: {
        slug,
      },
    })
  })
}
```

## Complex types

### Image

In sensenet images can be references. In this case the LeadImage field of a Blogpost will be something like this:

```json
"LeadImage": {
  "__deferred": {
    "uri": "/odata.svc/Root/Content/SampleWorkspace/Blog('2021-01-13-docviewer-updates')/LeadImage"
  }
},
```

Reference fields have to be expanded. You can define the fields you would like expand in the gatsby-config by listing the fieldNames in an array, eg.:
`oDataOptions: {expand: ['LeadImage']},`

> Then you can create a remote file node from it:

Reviewer: ```suggestion
Then you can create a remote file node from the returned data:
```

```javascript
exports.onCreateNode = async ({ node, actions: { createNode }, createNodeId, getCache }) => {
  if (node.internal.type === 'sensenetBlogPost') {
    const leadImageNode = await createRemoteFileNode({
      url: `<YOUR REPOSITORY URL>${node.LeadImage.Path}`,
      parentNodeId: node.Id.toString(),
      createNode,
      createNodeId,
      getCache,
    })
    if (leadImageNode) {
      node.leadImage___NODE = leadImageNode.id ///connect to blogpost node
    }
  }
}
```

> <note severity="info">The content should be public otherwise you have to provide a httpHeaders: { Authorization: `Bearer <yourAccessToken>` } attribute to createRemoteFileNode function</note>

Reviewer: ```suggestion
<note severity="info">If you want to request content from a sensenet repository, they should be public otherwise you have to provide a httpHeaders: { Authorization: `Bearer <yourAccessToken>` } attribute to createRemoteFileNode function</note>
```
